### DISCUSSION TITLE
How do I attach bpf programs to perf events?

### DISCUSSION THREAD

**Discussion by Jun10ng** [cilium/ebpf#800] [Q&A] [answered]
Hi masters,

I want to pprof my prog by ebpf, so I need to call the interface `bpf_program__attach_perf_event` that I found on libbpf.

But cilium/ebpf seems not to implement the interface. 

Is it that I did not find this method? Or is there another way to achieve my goal?

**Comment by florianl** [comment, marked as answer]
Hi @Jun10ng 

In general `ebpf` and `perf` are two very different subsystem of the Linux Kernel. And as far as I can tell, libbpf only provides a very limited part of the `perf` subsystem to use with `ebpf`.

If you check out [github.com/cilium/ebpf/link](https://pkg.go.dev/github.com/cilium/ebpf@v0.9.2/link) you will notice, the package `link` make use of `perf` elements for its use in `Tracepoint`s and `Kprobe`s. This approaches are event based. So your eBPF program will only be triggered if the tracepoint or kprobe is used.

A different part of the `perf` subsystem is based on sampling. And I think you might look for a solution in this space. Implementing a more wider spectrum of the `perf` Kernel subsystem in Go is quite complex. But there are packages like [github.com/hodgesds/perf-utils](https://pkg.go.dev/github.com/hodgesds/perf-utils) and [github.com/acln0/perf](https://github.com/acln0/perf) that are doing a great job. After setting up the desired perf event for the sampling, you can just use the returned FD and attach the eBPF program to it via 
 `unix.IoctlSetInt(event.FD(), unix.PERF_EVENT_IOC_SET_BPF, prog.FD())`.

[Here](https://gist.github.com/florianl/5d9cc9dbb3822e03f6f65a073ffbedbb) I'm providing a minimal example of setting up a sample based profiling.
